app/post.go

```diff
@@ -1297,10 +1342,102 @@ func (a *App) convertUserNameToUserIds(usernames []string) []string {
 	return usernames
 }
 
+// GetLastAccessiblePostTime returns CreateAt time of last accessible post as per the cloud limit
+func (a *App) GetLastAccessiblePostTime(useCache bool) (int64, *model.AppError) {
+	cfg := a.Config()
+	if cfg.FeatureFlags == nil || !cfg.FeatureFlags.CloudFree {
+		return 0, nil
+	}
+
+	if useCache {
+		system, err := a.Srv().Store.System().GetByName(model.SystemLastAccessiblePostTime)
+		if err != nil {
+			var nfErr *store.ErrNotFound
+			switch {
+			case errors.As(err, &nfErr):
+				// All posts are accessible
+				return 0, nil
+			default:
+				mlog.Warn("GetLastAccessiblePostTime: failed to fetch from cache.", mlog.Err(err))
+			}
+		} else {
+			lastAccessiblePostTime, err := strconv.ParseInt(system.Value, 10, 64)
+			if err != nil {
+				return 0, model.NewAppError("GetLastAccessiblePostTime", "parse.int64", map[string]interface{}{"Value": system.Value}, err.Error(), http.StatusInternalServerError)
+			}
+
+			return lastAccessiblePostTime, nil
+		}
+	}
+
+	limits, err := a.Cloud().GetCloudLimits("")
+	if err != nil {
+		return 0, model.NewAppError("GetLastAccessiblePostTime", "api.cloud.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	if limits == nil || limits.Messages == nil || limits.Messages.History == nil {
+		// Cloud limit is not applicable
+		return 0, nil
+	}
+
+	createdAt, err := a.Srv().GetStore().Post().GetRecentNthPostTime(int64(*limits.Messages.History))
+	if err != nil {
+		return 0, model.NewAppError("GetLastAccessiblePostTime", "app.last_accessible_post.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	// Update Cache
+	err = a.Srv().Store.System().SaveOrUpdate(&model.System{
+		Name:  model.SystemLastAccessiblePostTime,
+		Value: strconv.FormatInt(createdAt, 10),
+	})
+	if err != nil {
+		mlog.Warn("GetLastAccessiblePostTime: failed to update cache.", mlog.Err(err))
+	}
+
+	return createdAt, nil
+}
+
+// filterInaccessiblePosts filters out the posts, past the cloud limit
+func (a *App) filterInaccessiblePosts(postList *model.PostList) *model.AppError {
+	if postList == nil || postList.Posts == nil || len(postList.Posts) == 0 {
```

Comment:
The main issue I see with this function is posts can be in time ascending or descending order, and this function needs to support both. In addition, while reviewing it, I realized it traverses the entire post list, but as far as I can tell we are guaranteed that the list is sorted either ascending or descending, so I think we can do a binary-search like operation. Since this is a fairly main path of the application, I think its worth optimizing for it.

I made a branch off yours that implements this search: https://github.com/mattermost/mattermost-server/pull/20439. What do you think?

Reply:
Thanks for the PR @neallred.

However, I noticed a couple of things:
1. This func `getPostsAround` doesn't seem to return posts ordered by CreatedAt, so we may still need to support the linear search.

```go
query = query.From(table).
		Where(conditions).
		// Adding ChannelId and DeleteAt order columns
		// to let mysql choose the "idx_posts_channel_id_delete_at_create_at" index always.
		// See MM-24170.
		OrderBy("p.ChannelId", "p.DeleteAt", "p.CreateAt "+sort).
		Limit(uint64(options.PerPage)).
		Offset(uint64(offset))
```

2. There are a few other funcs as well like `GetParentsForExportAfter` which don't return ordered posts. But those are used by app/exports.go, although not sure, if these need to be limited as well.

Reply:
> 1. ... getPostsAround doesn't ... return posts ordered by CreatedAt, so we may still need to support the linear search.

I missed that one, I think you're right. In that case I'd leave it up to others more familiar with this code whether we want to somehow support both the linear path as well as an optimized path for lists that are sorted by CreatedAt. Could be an extra boolean parameter or `filterOptions` struct parameter, that lets us add more optimizations to the limiting in the future.

> 2. ... other funcs ... like GetParentsForExportAfter ... used by app/exports.go

I don't think those should be limited. [Prior internal discussion](https://community.mattermost.com/private-core/pl/f9npw3nprbrszpe5id1wt9nuno) was that message export should always be unlimited so that we don't lock people into our platform (one of the reasons Mattermost was created in the first place: frustration with vendor lock-in).

Reply:
I merged the PR and attempted to sort out everywhere that looked like it was sorted by CreateAt vs not. Went the route of giving `filterInaccessiblePosts` an options struct parameter so that the caller can choose what searching technique is appropriate.